concurrency_limit/context_managers.py

```diff
@@ -1,78 +1,94 @@
+        limit_configuration = LimitConfiguration(key='my_key', limit=5, limit_timeout=10, limit_expire=30)
+
+        with limit(redis_configuration, limit_configuration) as count:
+            # Scoped block of code that will be executed under the concurrency limit.
+            print(f"Executing the scoped block of code. Current count: {count}")
+
+    The method acquires an execution slot by increasing a concurrency counter stored in Redis, and checks if the
+    counter is below the configured limit. If the limit is exceeded, the method waits for the configured interval
+    before trying again. If the configured timeout is reached, a `ConcurrencyLimitExceededException` is raised. Upon
+    exiting the scoped block, the context manager releases the execution slot and updates the concurrency counter
+    in Redis accordingly.
+
+    Note: Any exceptions raised within the scoped block will propagate outside the scope of the `limit` method.
+
+    :param redis_configuration: RedisConfiguration object containing the configuration details for connecting to Redis.
+    :param limit_configuration: LimitConfiguration object containing the configuration details for the limit.
     """
+
+    class _LockAcquireException(Exception):
+        pass
+
     client = get_redis(redis_configuration)
-    start = time.time()
-    executed = False
+    start = time.monotonic()
+
+    lock_pattern = f"{limit_configuration.key}:*"
+    lock_key = (
+        f"{limit_configuration.key}:{time.monotonic_ns()}-{random.randint(1000, 9999)}"
+    )
 
     try:
-        # we loop as long as it was not possible to execute the context manager's scope.
-        while not executed:
-            # first we try to acquire an execution slot by increasing the concurrency counter and checking if the
-            # counter is below the configured limit. if it is, we execute the scope. if it is not, we immediately
-            # decrease the concurrency counter as we are not able to execute the scope.
-            count = client.incr(limit_configuration.key, 1)
-            client.expire(limit_configuration.key, limit_configuration.limit_expire)
-
-            # if there is no available execution slot.
-            if count > limit_configuration.limit:
-                client.decr(limit_configuration.key, 1)
-                elapsed = time.time() - start
-
-                # if we are waiting longer than the configured timeout, we raise a `ConcurrencyLimitExceededException`
-                # exception. executing the context manager's scope failed in this case.
+        # We loop as long as it was not possible to execute the context manager's scope.
+        while True:
+            try:
+                # First we check if we should try to acquire an execution slot. If the limit is already exceeded,
+                # there is no need to set the lock-key.
+                count = len(client.keys(lock_pattern))
```

Comment:
Using `KEYS` on Redis is not recommended for production use as it blocks the whole DB until all entries are iterated. `SCAN` should always be preferred!
I don't like the performance implication of either method for this. Large amounts of limits and entries will cause issues. I would suggest using a Redis `Set` containing actual timestamps with normal expiration (only for key cleanup). Currently running executions can be checked with `SCARD`. In case we hit the limit, we should get all members with `SMEMBERS` and check if there are any outdated ones and remove them. Maybe adding a key with the timestamp of the last sanity check would make sense to prevent unecessary cleanups.

Reply:
I've changed the implementation similar to your suggestion, but using hashes.